File: app/lib/models/photo.dart
/// Photo data models.
class Photo {
  final String id;
  final String userId;
  final String? uploadedBy;
  final String fileHash;
  final int fileSize;
  final String mimeType;
  final int? width;
  final int? height;
  final String? takenAt;
  final double? latitude;
  final double? longitude;
  final String? locationName;
  final String? cameraMake;
  final String? cameraModel;
  final String? description;
  final bool isFavorite;
  final bool isVideo;
  final double? duration;
  final String createdAt;
  final String thumbSmallUrl;
  final String thumbMediumUrl;

  const Photo({
    required this.id,
    required this.userId,
    this.uploadedBy,
    required this.fileHash,
    required this.fileSize,
    required this.mimeType,
    this.width,
    this.height,
    this.takenAt,
    this.latitude,
    this.longitude,
    this.locationName,
    this.cameraMake,
    this.cameraModel,
    this.description,
    required this.isFavorite,
    required this.isVideo,
    this.duration,
    required this.createdAt,
    required this.thumbSmallUrl,
    required this.thumbMediumUrl,
  });

  factory Photo.fromJson(Map<String, dynamic> json, {String serverUrl = ''}) {
    return Photo(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      uploadedBy: json['uploaded_by'] as String?,
      fileHash: json['file_hash'] as String,
      fileSize: json['file_size'] as int,
      mimeType: json['mime_type'] as String,
      width: json['width'] as int?,
      height: json['height'] as int?,
      takenAt: json['taken_at'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      locationName: json['location_name'] as String?,
      cameraMake: json['camera_make'] as String?,
      cameraModel: json['camera_model'] as String?,
      description: json['description'] as String?,
      isFavorite: json['is_favorite'] as bool? ?? false,
      isVideo: json['is_video'] as bool? ?? false,
      duration: (json['duration'] as num?)?.toDouble(),
      createdAt: json['created_at'] as String,
      thumbSmallUrl: '$serverUrl${json['thumb_small_url']}',
      thumbMediumUrl: '$serverUrl${json['thumb_medium_url']}',
    );
  }

  /// Parse takenAt or createdAt into a DateTime for grouping.
  DateTime get displayDate {
    final src = takenAt ?? createdAt;
    return DateTime.tryParse(src) ?? DateTime.now();
  }
}

class PhotoList {
  final List<Photo> photos;
  final String? nextCursor;
  final int totalCount;

  const PhotoList({
    required this.photos,
    this.nextCursor,
    required this.totalCount,
  });

  factory PhotoList.fromJson(Map<String, dynamic> json, {String serverUrl = ''}) {
    return PhotoList(
      photos: (json['photos'] as List).map((e) => Photo.fromJson(e, serverUrl: serverUrl)).toList(),
      nextCursor: json['next_cursor'] as String?,
      totalCount: json['total_count'] as int,
    );
  }
}

File: server/schemas/photo.py
"""Photo request/response schemas."""

from typing import Optional

from pydantic import BaseModel


class PhotoCheckRequest(BaseModel):
    hashes: list[str]


class PhotoCheckResponse(BaseModel):
    existing: list[str]
    new: list[str]


class PhotoUploadResponse(BaseModel):
    photo_id: str
    thumb_url: str
    status: str


class PhotoResponse(BaseModel):
    id: str
    user_id: str
    uploaded_by: Optional[str] = None
    file_hash: str
    file_size: int
    mime_type: str
    width: Optional[int]
    height: Optional[int]
    taken_at: Optional[str]
    latitude: Optional[float]
    longitude: Optional[float]
    location_name: Optional[str]
    camera_make: Optional[str]
    camera_model: Optional[str]
    ai_scene: Optional[str] = None
    exif_data: Optional[str] = None
    description: Optional[str]
    is_favorite: bool
    is_video: bool
    duration: Optional[float]
    created_at: str
    thumb_small_url: str
    thumb_medium_url: str


class PhotoListResponse(BaseModel):
    photos: list[PhotoResponse]
    next_cursor: Optional[str]
    total_count: int


class PhotoUpdateRequest(BaseModel):
    is_favorite: Optional[bool] = None
    location_name: Optional[str] = None
    description: Optional[str] = None


class PhotoBatchRequest(BaseModel):
    action: str  # 'delete' | 'favorite' | 'unfavorite'
    photo_ids: list[str]


class PhotoBatchResponse(BaseModel):
    success_count: int
    failed_count: int


class SystemStatusResponse(BaseModel):
    server_name: str
    server_id: str
    server_url: str
    version: str
    photo_count: int
    total_size_bytes: int
    storage_total_bytes: int
    storage_used_bytes: int
    storage_free_bytes: int
    storage_usage_percent: float
